### ISSUE TITLE
get_path(id) fails if file changed on disk?

### ISSUE THREAD

**Issue by davidbrochart** [jupyter-server/jupyter_server_fileid#15]
## Description

It looks like if a file has modifications, `get_path(id)` will return `None` and delete the entry for this file in the database. At least this is how I interpret [this line](https://github.com/jupyter-server/jupyter_server_fileid/blob/7f92c97bc384828a2c712be1d1644315645bb6ba/jupyter_server_fileid/manager.py#L238).
Could you confirm @dlqqq ?

## Reproduce

I can see that in https://github.com/jupyterlab/jupyterlab/pull/13246 after opening a document in collaborative mode and making a change to the document.

**Comment by dlqqq** [comment]
@davidbrochart Unfortunately, this is a quirk that is present on Linux. Essentially, the `stat()` system call exposed by the Linux kernel doesn't expose file creation timestamps as that was not supported by the ext* filesystems at the time this system call was written. Ext4 supports them, but this was placed into a separate system call `statx()` (presumably "extended" stat). 

As a result, we have no way of reliably distinguishing between a file that was edited v.s. a file that was deleted and created at the same path. Consider a file `Untitled.ipynb`. These two user flows are entirely indistinguishable with the information from `stat()`:

1. delete `Untitled.ipynb`, create a new empty `Untitled.ipynb`
2. edit `Untitled.ipynb` with some new content

So, by default, the File ID manager is as strict (pessimistic) as possible. However, FileIdManager can listen to "save" events emitted by the ContentsManager, which means we can simply update the modified timestamps whenever an edit is performed from the JupyterLab UI (in-band edit). This also would require JS2 to be installed, unless [we backport events to JS1](https://github.com/jupyter-server/jupyter_server/issues/960) (which we closed b/c JS2 will be the default in JL 3.5). I had intended but forgot to include this in #8 unfortunately.

However, this would still mean edits done outside JupyterLab (out-of-band edits, e.g. via `vim` in the terminal) are untrackable so far. To handle out-of-band edits, we would need creation timestamps which will not be available for filesystems on ext3 and under. And to get creation timestamps, we need to call the `statx()` system call from Python, which unfortunately seems to require us to use the `ctypes` API to directly interface with C.

Keep in mind that this behavior should work fine on Windows and Mac OS, since both support file creation timestamps on their filesystems (NTFS, HFS/HFS+/APFS resp.).

In summary:
- edits create new file IDs by default in Linux
- tracking in-band edits requires listening to JS2 events (working on this right now)
- tracking out-of-band edits requires ext4 filesystem and `statx()` call

**Comment by kevin-bates** [comment]
>- edits create new file IDs by default in Linux

How are the external FileID references updated in these "edited on Linux" scenarios?

To help distinguish between delete/create vs. update can you use `st_ino` and, if equal, consider it an update?  Or are you finding that inodes are immediately re-used (on the filesystems you've checked)?

**Comment by dlqqq** [comment]
> How are the external FileID references updated in these "edited on Linux" scenarios?

They're not. File IDs are never updated by design; a different file ID means that it should be considered an entirely distinct file.

> To help distinguish between delete/create vs. update can you use st_ino and, if equal, consider it an update? Or are you finding that inodes are immediately re-used (on the filesystems you've checked)?

Ext4 and APFS both consistently reuse inode numbers after deleting and immediately creating a new file. I believe that's sufficient evidence to demonstrate that a "both path and inode number are equal" check is insufficient.

I'm not exactly keen on adding a configurable trait that allows users to lower the "strictness" of the file ID manager (e.g. assume delete/creates at same path to be updates), given that both JS2 allows us to distinguish edits, and that there is a plan for `statx()` on ext4 filesystems. To get this bug in the future, a user will need to both be using JS1 and be using an older filesystem like ext3.

**Comment by dlqqq** [comment]
Edits now work so long as they're done via the ContentsManager API in JS2.

- Supporting JS1 requires backporting events to JS1
- Supporting out-of-band edits requires `statx()` system call integration

**Comment by davidbrochart** [comment]
Thanks, I've checked that it works fine in https://github.com/jupyterlab/jupyterlab/pull/13246.

**Comment by dlqqq** [comment]
I'll still need to cut a release that includes this change; I was going to cut a release after getting a performance enhancement for `get_path()`.

**Comment by dlqqq** [comment]
LMK if you need it sooner.

**Comment by davidbrochart** [comment]
It's already [done](https://github.com/jupyter-server/jupyter_server_fileid/releases/tag/v0.3.1). Let's only use [jupyter-releaser](https://github.com/jupyter-server/jupyter_releaser) from now on.
